File: 基础课/动态规划/背包/01-优化.cpp
#include <iostream>
using namespace std;

const int N = 1010;
int f[N]; // 表示在 j 容量时，能够得到的价值的最大值
int w[N], v[N];
int n, m;

int main()
{

  cin >> n >> m;

  for (int i = 1; i <= n; i++)
    cin >> v[i] >> w[i];

  for (int i = 1; i <= n; i++)
    for (int j = m; j >= v[i]; j--)
    {
      f[j] = max(f[j], f[j - v[i]] + w[i]);
    }

  cout << f[m] << endl;

  return 0;
}

File: 基础课/动态规划/背包/py/01优化.py
def knapsack(weights, values, capacity):
    n = len(weights)
    dp = [0 for _ in range(capacity + 1)]
    for i in range(n):
        for j in range(capacity, weights[i] - 1, -1):
            dp[j] = max(dp[j], dp[j - weights[i]] + values[i])
    
    return dp[capacity]

# 生产单元测试, 使用 assert 语法
# 1. 成功用例
weights = [1, 3, 4, 5]
values = [1, 4, 5, 7]
capacity = 7

assert knapsack(weights, values, capacity) == 9